x/virtualgroup/keeper/msg_server.go

```diff
@@ -0,0 +1,187 @@
+package keeper
+
+import (
+	"context"
+
+	sdkerrors "cosmossdk.io/errors"
+	"cosmossdk.io/math"
+	sptypes "github.com/bnb-chain/greenfield/x/sp/types"
+	"github.com/bnb-chain/greenfield/x/virtualgroup/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+func (k msgServer) UpdateParams(goCtx context.Context, req *types.MsgUpdateParams) (*types.MsgUpdateParamsResponse, error) {
+	if k.GetAuthority() != req.Authority {
+		return nil, sdkerrors.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", k.GetAuthority(), req.Authority)
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if err := k.SetParams(ctx, req.Params); err != nil {
+		return nil, err
+	}
+
+	return &types.MsgUpdateParamsResponse{}, nil
+}
+
+func (k msgServer) CreateGlobalVirtualGroup(goCtx context.Context, req *types.MsgCreateGlobalVirtualGroup) (*types.MsgCreateGlobalVirtualGroupResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	spOperatorAddr := sdk.MustAccAddressFromHex(req.PrimarySpAddress)
+
+	sp, found := k.spKeeper.GetStorageProviderByOperatorAddr(ctx, spOperatorAddr)
+	if !found {
+		return nil, sptypes.ErrStorageProviderNotFound
+	}
+
+	var secondarySpIds []uint32
+	for _, id := range req.SecondarySpIds {
+		ssp, found := k.spKeeper.GetStorageProvider(ctx, id)
+		if !found {
+			return nil, sdkerrors.Wrapf(sptypes.ErrStorageProviderNotFound, "secondary sp not found, ID: %d", id)
+		}
+		secondarySpIds = append(secondarySpIds, ssp.Id)
+	}
+
+	gvgFamily, err := k.GetOrCreateEmptyGVGFamily(ctx, req.FamilyId, sp.Id)
+
+	if err != nil {
+		return nil, err
+	}
+
+	gvgID := k.GenNextGVGID(ctx)
+	if gvgID == 0 {
+		return nil, sdkerrors.Wrapf(types.ErrGenSequenceIDError, "wrong next gvg id.")
+	}
+
+	// deposit enough tokens for oncoming objects
+	coins := sdk.NewCoins(sdk.NewCoin(k.DepositDenomForGVG(ctx), req.Deposit.Amount))
+	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, sdk.MustAccAddressFromHex(sp.FundingAddress), types.ModuleName, coins)
+	if err != nil {
+		return nil, err
+	}
+
+	gvg := &types.GlobalVirtualGroup{
+		Id:                    k.GenNextGVGID(ctx),
```

Comment:
Use `gvgID`?

Reply:
It's a bit redundant for using `gvgID`

Reply:
I mean use `gvgID ` replace `k.GenNextGVGID(ctx)`. The `gvgID ` has been generated above.